### PR TITLE
i#2905: ASSERT prof_pcs on ARM

### LINKED ISSUE #2905: ASSERT -prof_pcs on ARM
When running a client which inserts any clean call with `-prof_pcs` on ARM, I see the following assertion:

```
(xenial)piazzt@localhost:/tmp/dynamorio-master$ ./bin32/drrun -debug -prof_pcs -c ./api/bin/libinscount.so -- true
<Starting application /bin/true (19351)>
...
Client inscount is running
<Application /bin/true (19351).  Internal Error: DynamoRIO debug check failure: Not implemented @/home/piazzt/SD/Build/dynamorio-master/core/arch/instr_shared.c:3088 (0)
(Error occurred @0 frags)
version 7.0.17614, custom build
-no_dynamic_options -prof_pcs -client_lib '/tmp/dynamorio-master/api/bin/libinscount.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct
/tmp/dynamorio-master/api/bin/libinscount.so=0xb493a000
/usr/lib/arm-linux-gnueabihf/libstdc++.so.6=0xb479a000
/lib/arm-linux-gnueabihf/libgcc_s.so.1=0xb4606000
/lib/arm-linux-gnueabihf/libm.so.6=0xb4721000
/lib/arm-linux-gnueabihf/libc.so.6=0xb4634000
/lib/ld-linux-armhf.so.3=0xb48f7000
/tmp/dynamorio-master/ext/lib32/debug/libdrmgr.so=0xb4921000>
```

### PR DESCRIPTION
**Pull request by toshipiazza**
Fixes an NYI on ARM when running -prof_pcs with a client that emits a clean
call by reusing code pre-existing with Aarch64

Fixes #2905

### CODE REVIEW

**Comment by toshipiazza** [comment]
Note: this fixes the bug as described in the original issue, but on a custom client I wrote I'm hitting a new assertion now. I don't think it's related but I'll do more investigation.

**Comment by fhahn** [comment]
The AArch32 bot fail, because I've update the compiler and we now get the `IT blocks containing ....` messages which are treated as errors.  I'll have another stab at #1675

File: core/arch/mangle_shared.c [hunk 1]
```diff
@@ -430,7 +430,7 @@ insert_meta_call_vargs(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
              * which specifies that the caller must save xax (and xcx and xdx)
              */
             insert_get_mcontext_base(dcontext, ilist, instr, SCRATCH_REG0);
```
Comment (toshipiazza) [inline core/arch/mangle_shared.c]: I realized that this is wrong. `SCRATCH_REG0` is `r0` on ARM, which is the first argument to each clean call... x86 is clearly fine though. I'm curious, how does this work for Aarch64 either then? Isn't `x0` the first argument to clean calls as well?

@derekbruening thoughts?

Comment (fhahn) [reply core/arch/mangle_shared.c]: I could not find a test that enables profile_pcs on AArch64. We should definitely add some. This looks wrong, as the clobbering of `r0`/`x0` happens after the parameters are moved to the registers. I think we could use the link register here on ARM and AArch64, as we use `blx` to branch to the function. If we only allow AAPCS64 calls on AArch64, we should be able to use any of  the temporary registers (r9...r15), which the caller has to save & restore. 

Also, if we limit clean calls to AAPCS64 complaint calls on AArch64, we could also avoid saving/restoring r19..r28?

Comment (fhahn) [reply core/arch/mangle_shared.c]: @egrimley may have some thoughts as well

Comment (toshipiazza) [reply core/arch/mangle_shared.c]: Using the link reg seems to work on ARM32, I'll update the PR accordingly. Thanks!

Comment (derekbruening) [reply core/arch/mangle_shared.c]: Sorry, I only evaluated whether ARM would be different from A64 and assumed the A64 code was correct without thinking about it much.

File: core/arch/mangle_shared.c [hunk 2]
```diff
@@ -426,29 +426,31 @@ insert_meta_call_vargs(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
...
         if (SCRATCH_ALWAYS_TLS()) {
-            /* SCRATCH_REG0 is dead here, because clean calls only support "cdecl",
-             * which specifies that the caller must save xax (and xcx and xdx)
-             */
-            insert_get_mcontext_base(dcontext, ilist, instr, SCRATCH_REG0);
-# ifdef AARCH64
+# ifdef AARCHXX
+            /* DR_REG_LR is dead here */
+            insert_get_mcontext_base(dcontext, ilist, instr, DR_REG_LR);
             /* TLS_REG1_SLOT is not safe since it may be used by clients.
```
Comment (fhahn) [inline core/arch/mangle_shared.c]: This comment needs updating I think

Comment (toshipiazza) [reply core/arch/mangle_shared.c]: Done.

**Comment by fhahn** [comment]
Could you also enable the  `code_api|sample.inscount.prof-pcs.cleancall ` test in `suite/tests/CMakeLists.txt` for ARM (around line 2371

Strangely enough, this test fails now on AArch64 Precommit. I tried your patch manually on the same machine and it passes.... I can have another look tomorrow.

**Comment by toshipiazza** [comment]
Done, added `code_api|sample.inscount.prof-pcs.cleancall` to the CMakeLists.txt, the test seems to work for me.

**Comment by toshipiazza** [comment]
One last question: when I run -prof_pcs with a custom client on a long-running ARM application, I see the following metrics:

```
DynamoRIO library: 0x71000000-0x71184000
client library: 0xadd6e000-0xadd85000
ITIMER distribution (185076):
    6.5% of time in INTERPRETER (11999)
    0.0% of time in DISPATCH (31)
    0.0% of time in SYSCALL HANDLER (3)
    2.0% of time in INDIRECT BRANCH LOOKUP (3639)
   15.5% of time in FRAGMENT CACHE (28726)

PC PROFILING RESULTS
```

These percentages don't add up to 100, is this a problem or just unusual?

**Comment by fhahn** [comment]
Could it be that something else is still missing? On my Raspberry Pi I get `Virtual timer expired` when using `-prof_pcs` with `-debug` and a segfault without `-debug`. `code_api|sample.inscount.prof-pcs.cleancall` fails on AArch32 precommit, probably for the same reason.

> These percentages don't add up to 100, is this a problem or just unusual?

I think all samples should be attributed to a category and there is an `UNKNOWN` bucket. That's just my understanding from a quick glance, @derekbruening probably knows for certain :)

**Comment by derekbruening** [comment]
> These percentages don't add up to 100, is this a problem or just unusual?

They should add up, so something is wrong.  Looking at pcprofile_results() it looks like it fails to print out the clean call bucket -- so I would guess that's where the rest of your are.  That should be added to the code.

**Comment by toshipiazza** [comment]
Looks like code_api|sample.inscount.prof-pcs.cleancall is still failing on the Aarch32 pre-commit, which I'm guessing is due to the race documented in #2907. Should I merge this in now or wait until the issue is resolved?

**Comment by fhahn** [comment]
If code_api|sample.inscount.prof-pcs.cleancall fails consistently with this change, we should either get to the bottom of #2907, do not enable the test just yet (but add a comment mentioning #2907) or mark it as flaky in runsuite_wrapper.pl, with a reference to #2907.

Obviously, fixing the issue to start with would be the best option :)

**Comment by toshipiazza** [comment]
Unfortunately I don't have so much time to look at this issue, so I've marked it as flaky for now.

**Comment by toshipiazza** [comment]
Hmm, x86 Linux failure for code_api|pthreads.ptsig doesn't look related. Same for ARM failure for code_api|linux.sigsuspend.

**Comment by fhahn** [comment]
I've created #2922 to mark sigsuspend as flaky on ARM. I'll get that in after the tests pass.

**Comment by toshipiazza** [comment]
Thanks @fhahn!

**Pull request merged**